### PR TITLE
Adding node selectors, affinity and tolerations to the dashboard pods

### PR DESCRIPTION
**Pull request by madhukarmmallia-plivo**

### CODE REVIEW

**Review by swoehrl-mw — COMMENTED** [review]
Hi @madhukarmmallia-plivo. Thank you for your PR. 
Can you add a little unittest to check the new fields are correctly added to the dashboard deployment (more as a regression test)? Probably best in `controllers/dashboard_test.go`. Other than that it looks good.

**Comment by madhukarmmallia-plivo** [comment]
Hello @swoehrl-mw! Could you please take a look now?

File: opensearch-operator/controllers/dashboard_test.go [hunk 1]
```diff
@@ -119,5 +119,30 @@ var _ = Describe("Dashboards Reconciler", func() {
...
+				Expect(deployment.Spec.Replicas).To(Equal(2))
+				Expect(deployment.Spec.Template.Spec.Containers[0].Resources.Limits.Cpu().String()).To(Equal("500m"))
+				Expect(deployment.Spec.Template.Spec.Containers[0].Resources.Limits.Memory().String()).To(Equal("1Gi"))
+				Expect(deployment.Spec.Template.Spec.Tolerations).To(ContainElement(corev1.Toleration{
+					Effect:   "NoSchedule",
+					Key:      "foo",
+					Operator: "Equal",
+					Value:    "bar",
+				}))
+				Expect(deployment.Spec.Template.Spec.NodeSelector).To(ContainElement(map[string]string{
```
Comment (swoehrl-mw) [inline opensearch-operator/controllers/dashboard_test.go:141]:
```suggestion
				Expect(deployment.Spec.Template.Spec.NodeSelector).Should(Equal(map[string]string{
```
The `ContainElement` does not work as intended for maps.

File: opensearch-operator/controllers/dashboard_test.go [hunk 2]
```diff
@@ -119,5 +119,30 @@ var _ = Describe("Dashboards Reconciler", func() {
...
+					"foo": "bar",
+				}))
+				Expect(deployment.Spec.Template.Spec.Affinity).To(ContainElement(corev1.Affinity{}))
```
Comment (swoehrl-mw) [inline opensearch-operator/controllers/dashboard_test.go:144]: This is not a list, so you need to use `Expect(...).To(Equal(...))`. And you are comparing pointer to value, so you need to dereference the pointer first.

File: opensearch-operator/controllers/dashboard_test.go [hunk 3]
```diff
@@ -119,5 +119,30 @@ var _ = Describe("Dashboards Reconciler", func() {
...
+			if OpensearchCluster.Spec.Dashboards.Enable {
+				dashboardDeployName := fmt.Sprintf("%s-dashboards", OpensearchCluster.Name)
+				deployment := &sts.Deployment{}
+				Eventually(func() error {
+					return k8sClient.Get(context.Background(), types.NamespacedName{
+						Name:      dashboardDeployName,
+						Namespace: OpensearchCluster.Namespace,
+					}, deployment)
+				}, timeout, interval).Should(Succeed())
+				Expect(deployment.Spec.Replicas).To(Equal(2))
```
Comment (swoehrl-mw) [inline opensearch-operator/controllers/dashboard_test.go:132]: You are comparing a pointer with a value, and the configured replicas is `3`. You can use the `k8s.io/utils/pointer` package and create a pointer:
```suggestion
				Expect(deployment.Spec.Replicas).To(Equal(pointer.Int32(3)))
```

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
There are some problems with the comparisons in your test, see my code comments. You can always run the tests locally with `make test` to make sure they work before pushing.

Regarding the failing lint job: Not sure why but looks like a out-of-memory kill. You can locally verify there are no lint problems by running `make lint`.

**Comment by madhukarmmallia-plivo** [comment]
@swoehrl-mw I have updated tests as per your feedback. Could you please review it?

**Comment by madhukarmmallia-plivo** [comment]
Hi @idanl21 Could you also please have a look at this?

**Review by swoehrl-mw — APPROVED** [review]
LGTM

**Pull request merged**